src/services/datasource.ts

```diff
@@ -25,26 +44,82 @@ class WrappedLokiDatasource extends RuntimeDataSource<DataQuery> {
       getDataSourceSrv()
         .get(getDataSource(request.scopedVars.__sceneObject.valueOf()))
         .then((ds) => {
-          // override the target datasource to Loki
-          request.targets = request.targets.map((target) => {
-            target.datasource = ds;
-            return target;
-          });
-
-          // query the datasource and return either observable or promise
-          const dsResponse = ds.query(request);
-          if (isObservable(dsResponse)) {
-            dsResponse.subscribe(subscriber);
-          } else {
-            dsResponse.then((response) => {
-              subscriber.next(response);
-              subscriber.complete();
-            });
+          if (!(ds instanceof DataSourceWithBackend)) {
+            throw new Error('Invalid datasource!');
+          }
+
+          const requestType = request.targets?.[0]?.resource;
+
+          switch (requestType) {
+            case 'volume': {
+              this.transformVolumeResponse(request, ds, subscriber);
+              break;
+            }
+            case 'patterns': {
+              console.warn('not yet implemented');
+              // this.transformPatternResponse(request, ds, subscriber);
+              break;
+            }
+            default: {
+              // override the target datasource to Loki
+              request.targets = request.targets.map((target) => {
+                target.datasource = ds;
+                return target;
+              });
+
+              // query the datasource and return either observable or promise
+              const dsResponse = ds.query(request);
+              dsResponse.subscribe(subscriber);
+              break;
+            }
           }
         });
     });
   }
 
+  private transformVolumeResponse(
+    request: DataQueryRequest<LokiQuery & SceneDataQueryResourceRequest>,
+    ds: DataSourceWithBackend<LokiQuery>,
+    subscriber: Subscriber<DataQueryResponse>
+  ) {
+    if (request.targets.length !== 1) {
+      throw new Error('Volume query can only have a single target!');
+    }
+
+    const targetsInterpolated = ds.interpolateVariablesInQueries(request.targets, request.scopedVars);
+
+    const dsResponse = ds.getResource(
+      'index/volume',
+      {
+        expr: targetsInterpolated[0].expr,
```

Comment:
```suggestion
```

I didn't find this in https://grafana.com/docs/loki/latest/reference/loki-http-api/#query-log-volume

Reply:
yeah this was a leftover from me playing around, good catch